client/starwhale/api/_impl/dataset/loader.py

```diff
@@ -92,6 +94,44 @@ def _do_iter_row(self) -> t.Generator[TabularDatasetRow, None, None]:
             for row in self.tabular_dataset.scan():
                 yield row
         else:
+
+            def _travel_link(obj: t.Any) -> t.List[Link]:
+                _lks = []
+                if isinstance(obj, Link):
+                    _lks.append(obj)
+                elif isinstance(obj, dict):
+                    for v in obj.values():
+                        _lks.extend(_travel_link(v))
+                elif isinstance(obj, (list, tuple)):
+                    for v in obj:
+                        _lks.extend(_travel_link(v))
+                elif isinstance(obj, SwObject):
+                    for v in obj.__dict__.values():
+                        _lks.extend(_travel_link(v))
+                return _lks
+
+            def sign_uris(uris: t.List[str]) -> str:
+                r = (
+                    CloudRequestMixed()
+                    .do_http_request(
+                        f"/project/{self.dataset_uri.project}/{self.dataset_uri.object.typ}/{self.dataset_uri.object.name}/version/{self.dataset_uri.object.version}/sign-links",
+                        method=HTTPMethod.POST,
+                        instance_uri=self.dataset_uri,
+                        params={
+                            "expTimeMillis": int(
+                                os.environ.get(
+                                    "SW_MODEL_PROCESS_UNIT_TIME_MILLIS", "60000"
+                                )
+                            )
+                            * self.session_consumption.batch_size,  # type: ignore
```

Comment:
does mypy cannot get `self.session_consumption.batch_size` type?

Reply:
self.session_consumption is None?

Reply:
https://github.com/star-whale/starwhale/pull/1516/files#diff-1b1cb0ab4ea118c71c67544145a896cf33b27e918cf8e76e3fa181aee4d1d372L31  self.session_consumption maybe None.

Reply:
In the `else` branch , `self.session_consumption` couldn't be None. The reason why mypy cannot get `self.session_consumption.batch_size` type is that `self.session_consumption` is declared as `TabularDatasetSessionConsumption` which is subclass of `Protocol` i guess. But not sure.
```python
    def _do_iter_row(self) -> t.Generator[TabularDatasetRow, None, None]:
        if not self.session_consumption:
            for row in self.tabular_dataset.scan():
                yield row
        else:
            while True:
                # TODO: multithread for get meta, get data and ppl process
                pk = [self.last_processed_range] if self.last_processed_range else None
                rt = self.session_consumption.get_scan_range(pk)
                self.last_processed_range = rt
                if rt is None:
                    break

                if self.dataset_uri.instance_type == InstanceType.CLOUD:
                    for rows in self.tabular_dataset.scan_batch(
                        rt[0], rt[1], self.session_consumption.batch_size  # type: ignore
                    ):
                        _links = []
                        for row in rows:
                            _links.extend([row.data_uri] + self._travel_link(row.annotations))
                        uri_dict = self.sign_uris([lk.uri for lk in _links])
                        for lk in _links:
                            lk.signed_uri = uri_dict.get(lk.uri, "")

                        for row in rows:
                            yield row
                else:
                    for row in self.tabular_dataset.scan(rt[0], rt[1]):
                        yield row
```